objects/src/accounts/code/mod.rs

```diff
@@ -91,6 +92,75 @@ impl AccountCode {
         })
     }
 
+    // TODO: Document.
+    pub fn from_components(
+        components: &[AccountComponent],
+        account_type: AccountType,
+    ) -> Result<Self, AccountError> {
+        let (merged_mast_forest, _) =
+            MastForest::merge(components.iter().map(|component| component.mast_forest()))
+                .map_err(|err| AccountError::AccountCodeMergeError(err.to_string()))?;
+
+        let mut procedures = Vec::new();
+        let mut proc_root_set = BTreeSet::new();
+        let mut component_storage_offset = if account_type.is_faucet() { 1 } else { 0 };
+        for component in components {
+            let AccountComponent { code: library, storage_slots } = component;
+            let component_storage_size: u8 = storage_slots
+                .len()
+                .try_into()
+                .map_err(|_| AccountError::StorageTooManySlots(storage_slots.len() as u64))?;
+
+            for module in library.module_infos() {
+                for proc_mast_root in module.procedure_digests() {
+                    // We cannot support procedures from multiple components with the same MAST root
+                    // since storage offsets/sizes are set per MAST root. Setting them again
+                    // for procedures where the offset has already been inserted would cause that
+                    // procedure of the earlier component to write to the wrong
+                    // slot.
+                    if !proc_root_set.insert(proc_mast_root) {
+                        return Err(AccountError::AccountCodeMergeError(format!(
+                            "procedure with MAST root {proc_mast_root} is present in multiple account components"
+                        )));
+                    }
+
+                    // Components that do not access storage need to have offset and size set to 0.
+                    let (storage_offset, storage_size) = if component_storage_size == 0 {
+                        (0, 0)
+                    } else {
+                        (component_storage_offset, component_storage_size)
+                    };
+
+                    // Note: Offset and size are validated in `AccountProcedureInfo::new`.
+                    procedures.push(AccountProcedureInfo::new(
+                        proc_mast_root,
+                        storage_offset,
+                        storage_size,
+                    )?);
+                }
+            }
+
+            component_storage_offset = component_storage_offset.checked_add(component_storage_size)
+              .expect("account procedure info constructor should return an error if the addition overflows");
```

Comment:
I modified the size check in `AccountProcedureInfo::new` because it would cause overflows in some cases: https://github.com/0xPolygonMiden/miden-base/pull/941/commits/3837e7522c5bb2279e01e8f6fb3a7c2674d86bdc.

If the check we need to do is indeed whether offset+size > 255, then I think it's correct now. If offset = 255 and size = 0, then this would error in the same method earlier so we don't need to handle it explicitly.

But I'm actually wondering if we really need to check that or if offset+size >= 255 would be correct?
The [maximum storage slot index is 254](https://github.com/0xPolygonMiden/miden-base/blob/b575e8ab12552559016a24281a7ef68e5ef943c9/miden-lib/asm/kernels/transaction/lib/account.masm#L104), right? Unless this is a leftover from an earlier version. Assuming it is correct though, if we have an offset of 200 and a size of 55, then that check ([here](https://github.com/0xPolygonMiden/miden-base/blob/b575e8ab12552559016a24281a7ef68e5ef943c9/objects/src/accounts/code/procedure.rs#L58)) would be `255 > 255 = false` and it would not error and allow access to slot 255. But shouldn't it only allow access up to and including 254 if that is indeed the maximum index? The [apply_storage_offset](https://github.com/0xPolygonMiden/miden-base/blob/b575e8ab12552559016a24281a7ef68e5ef943c9/miden-lib/asm/kernels/transaction/lib/account.masm#L360) function also uses `gt`, so perhaps I'm missing something else here.

Reply:
> But something like `offset=200` and `size=55` should still be valid, right?

Ah, yes. I think I confused myself. offset+size is already one larger than the maximum allowed index, so using > rather than >= is correct.

I double checked `apply_storage_offset` and it's correct. It panics if this statement evaluates to `false`: `offset + size > offset + index` where `index` is the one passed to `get_item` and friends. So `254 + 1 > 254 + 0 = true` (no panic) and `254 + 1 > 254 + 1 = false` (panic).